### ISSUE TITLE
[Question] Calculating ELBO

### ISSUE THREAD

**Issue by danielkelshaw** [SebFar/radial_bnn#1]
First of all I just wanted to say that I found this paper fascinating and thanks for releasing the code!

I have recently been working on implementing the [Weight Uncertainty in Neural Networks](https://arxiv.org/pdf/1505.05424.pdf) paper and I noticed that the calculation of the `ELBO` term is a little different to what you have implemented here.

*Equation 2* of the *WUINN* paper aims to approximate the loss by using the difference between the log probabilities of the prior and posterior distribution whereas your code implements a more direct approach to the KL Divergence by calculating the `entropy` and `cross entropy`.

You have noted in the comments that:

`entropy = \sum_i log \sigma_i + c`

I was wondering how you are able to use this approach rather than sampling from the posterior distribution? Is this a case of using the *reparameterization trick*?

I'd be grateful for any guidance!

Thanks again.

**Comment by SebFar** [comment]
Hi Daniel,

WUNN does the same thing we do, just with a different approximate posterior. The entropy of the prior distribution is just the expectation of the log probability density function of the prior on the approximate posterior distribution, and cross entropy the expectation of the log probability of the approximate posterior.

Where we calculate the entropy, we know the analytical form of the prior function and the approximate posterior function and in this case we can calculate the exact value, which is given by that equation (full derivation is in the Appendix of the paper, and is pretty involved). For the other term we can't do it analytically and need to sample from the approx posterior.

Hope that helps!

**Comment by danielkelshaw** [comment]
Thank you for your response!

Just to clarify: the `entropy` is the *expected* value of the log likelihood of the sample evaluated over the posterior distribution - and the `cross entropy` is the *expected* value of the log likelihood of the samples from the posterior under the prior?

Am I right in saying that as this is often intractable (not in this case) one of the standard approaches is to use `MFVI` to calculate the `ELBO` term?

I had a look at the Appendix but that derivation was rather daunting - I'll have to set aside some more time to look through that.

**Comment by SebFar** [comment]
Right, specifically we're talking about the entropy of the approximate posterior (under the approximate posterior distribution) and the cross-entropy between the approximate posterior and the prior. If you write it out, this rearranges to just be the KL between the approximate posterior and the prior.

Whether this is analytically tractable depends on your choices for the prior and approximate posterior. When you do MFVI, it is analytically tractable because everything is a Gaussian. In our case, one term is analytically tractable, but the other is not, so we do Monte Carlo integration to estimate the expectation.

Yes the full derivation is daunting. Once the transformation used is clear the rest is just algebra, if complicated algebra.

**Comment by danielkelshaw** [comment]
Thanks for your help, that was a very clear explanation!
